pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+          case Leaf(value):
+              handle(value)
+
+The notion of handling recursive data structures with pattern matching
+immediately gave rise to the idea of handling general recursive patterns
+with pattern matching.  Pattern matching would thus also be used to define
+recursive functions such as::
+
+  def fib(arg):
+      match arg:
+          case 0:
+              return 1
+          case 1:
+              return 1
+          case n:
+              return fib(n-1) + fib(n-2)
+
+As pattern matching was repeatedly integrated into new and emerging
+programming languages, its syntax slightly evolved and expanded.  The two
+first cases in the ``fib`` example above could be written more succinctly
+as ``case 0 | 1:`` with ``|`` denoting alternative patterns.  Moreover, the
+underscore ``_`` was generally accepted as a wildcard, a filler where neither
+the structure nor value of parts of a pattern were of substance.  Since the
+underscore is already frequently used in equivalent capacity in Python's
+iterable unpacking (e.g., ``_, _, third, _* = something``) we kept these
+universal standards.
+
+It is noteworthy that the concept of pattern matching has always been
+closely linked to the concept of functions.  The different case clauses
+have always been considered as something like semi-indepedent functions
+where pattern variables take on the role of parameters.  This becomes
+most apparent when pattern matching is written as an overloaded function,
+along the lines of::
+
+  def fib( 0 | 1 ):
+      return 1
+  def fib( n ):
+      return fib(n-1) + fib(n-2)
+
+Even though such a strict separation of case clauses into independent
+functions does not make sense in Python, we find that patterns share many
+syntactic rules with parameters, such as binding arguments to local
+variables only or that variable/parameter names must not be repeated for
+a particular pattern/function.
+
+With its emphasis on abstraction and encapsulation, object-oriented
+programming posed a serious challenge to pattern matching.  In short: in
+object-oriented programming, we can no longer view objects as tagged tuples.
+The arguments passed into the constructor do not necessarily specify the
+attributes or fields of the objects.  Moreover, there is no longer a strict
+ordering of an object's fields and some of the fields might be private and
+thus inaccessible.  And on top of this, the given object might actually be
+an instance of a subclass with slightly different structure.
+
+To address this challenge, patterns became increasingly independent of the
+original tuple constructors.  In a pattern like ``Node(left, right)``,
+``Node`` is no longer a passive tag, but rather a function that can actively
+check for any given object whether it has the right structure and extract a
+``left`` and ``right`` field.  In other words: the ``Node``-tag becomes a
+function that transforms an object into a tuple or returns ``None`` to
+indicate that it is not possible.
```

Comment:
I'm not sure we need to bring up returning None on failure (conceivably, the API could also raise an exception, since in either case the match machinery has to check for failure). Maybe "or returns some failure indicator if it is not possible"? "Failure indicator" is vague enough that raising an exception could be substituted.

Reply:
Agreed.